File: concepts/03_errors_and_resources/03_error_chaining_suppression_and_aggregation/ruby/test_02_multiple_cleanup_failures.rb
# 共同问题：body 与多个 cleanup 同时失败时怎样保留信息。
# 输入：body 失败和两层 ensure 失败；观察：外层失败占优，较早失败沿 cause 链保留。
# polyglot-family: errors_and_resources
# polyglot-concept: error_chaining_suppression_and_aggregation
# polyglot-related: languages/ruby/language/08_exceptions_resources_and_loading/
# polyglot-related+: test_058_exception_cause_and_backtrace.rb

require "assertions"

A = PolyglotAssertions

final = begin
  begin
    begin
      raise RuntimeError, "body"
    ensure
      raise ArgumentError, "inner cleanup"
    end
  ensure
    raise IOError, "outer cleanup"
  end
rescue => error
  error
end

A.same(IOError, final.class)
A.equal("outer cleanup", final.message)
A.same(ArgumentError, final.cause.class)
A.equal("inner cleanup", final.cause.message)
A.same(RuntimeError, final.cause.cause.class)
A.equal("body", final.cause.cause.message)

A.done

File: concepts/03_errors_and_resources/03_error_chaining_suppression_and_aggregation/lua/test_02_multiple_cleanup_failures.lua
-- Common question: how are several cleanup failures represented?
-- Inputs: a body failure and two to-be-closed resources that both fail.
-- Observations: reverse cleanup order, current-error handoff, and explicit aggregation.
-- polyglot-family: errors_and_resources
-- polyglot-concept: error_chaining_suppression_and_aggregation
-- polyglot-related: languages/lua/language/07_errors_and_resources/
-- polyglot-related+: test_052_close_order_unwind_and_failure_precedence.lua

local t = require("support.assertions")
local events = {}
local function failing(name)
    return setmetatable({}, {
        __close = function(_, current)
            events[#events + 1] = {name = name, current = current}
            error({kind = "close", name = name, cause = current}, 0)
        end,
    })
end

local ok, final = pcall(function()
    local first <close> = failing("first")
    local second <close> = failing("second")
    t.truth(first and second)
    error({kind = "body"}, 0)
end)
t.falsey(ok)
t.equal(events[1].name, "second")
t.equal(events[1].current.kind, "body")
t.equal(events[2].name, "first")
t.equal(events[2].current.name, "second")
t.equal(final.name, "first")
t.equal(final.cause.name, "second")

t.done()
